File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

       ListNode*reverse(ListNode*&head){
        if(head==NULL || head->next==NULL){
            return head;
        }
        ListNode* NewHead=reverse(head->next);
        ListNode* front = head->next;
        front->next=head;
        head->next=NULL;
        return NewHead;
    }
    ListNode* reverseList(ListNode* head) {
        // if(head==NULL || head->next==NULL){
        //     return head;
        // }
        // ListNode*temp = head;
        // ListNode* prev = NULL;
        // ListNode* forward = head;
        // while(temp){
        //     forward = temp->next;
        //     temp->next = prev;
        //     prev = temp;
        //     temp = forward;
        // }
        // return prev;
        return reverse(head);
        // return ans;
    }
};

File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode curr = head;
        ListNode prev = null;
        if(curr == null){
            return null;
        }
        if(curr.next == null){
            return head;
        }
        while(curr != null){
            ListNode forward = curr.next;
            curr.next = prev;
            prev = curr;
            curr = forward;
        }
        return prev;
    }
}